File: src/main/java/io/github/addoncommunity/galactifun/api/universe/StarSystem.java
package io.github.addoncommunity.galactifun.api.universe;

import javax.annotation.Nonnull;

import org.bukkit.inventory.ItemStack;

import io.github.addoncommunity.galactifun.api.universe.attributes.Orbit;
import io.github.addoncommunity.galactifun.api.universe.types.StarSystemType;

/**
 * A star system filled with celestial objects
 *
 * @author Mooy1
 */
public final class StarSystem extends UniversalObject {

    public StarSystem(String name, StarSystemType type, Orbit orbit, Galaxy orbiting, ItemStack baseItem) {
        super(name, type, orbit, orbiting, baseItem);
    }

    @Override
    public double distanceTo(@Nonnull UniversalObject other) {
        if (this.orbitLevel > other.orbitLevel) {
            return other.orbit().currentDistance() + distanceTo(other.orbiting());
        } else if (this.orbitLevel == other.orbitLevel) {
            return Math.abs(this.orbit().currentDistance() - other.orbit().currentDistance());
        } else {
            return super.distanceTo(other);
        }
    }

}

File: src/main/kotlin/io/github/addoncommunity/galactifun/api/objects/StarSystem.kt
package io.github.addoncommunity.galactifun.api.objects

import io.github.addoncommunity.galactifun.api.objects.properties.Orbit
import org.bukkit.Material
import kotlin.math.abs

class StarSystem(name: String, orbiting: UniversalObject, orbit: Orbit) :
    UniversalObject(name, Material.SUNFLOWER, orbiting, orbit) {

    override fun distanceTo(other: UniversalObject): Double {
        return if (orbitLevel > other.orbitLevel) {
            other.orbit.distance + distanceTo(other.orbiting)
        } else if (orbitLevel == other.orbitLevel) {
            abs(this.orbit.distance - other.orbit.distance)
        } else {
            super.distanceTo(other)
        }
    }
}